rust/ql/lib/codeql/rust/internal/TypeInference.qll

```diff
@@ -998,6 +1015,29 @@ private StructType inferLiteralType(LiteralExpr le) {
   )
 }
 
+pragma[nomagic]
+private AssociatedTypeTypeParameter getFutureOutputTypeParameter() {
+  result.getTypeAlias() = any(FutureTrait ft).getOutputType()
+}
+
+pragma[nomagic]
+private Type inferAwaitExprType(AwaitExpr ae, TypePath path) {
+  exists(TypePath exprPath | result = inferType(ae.getExpr(), exprPath) |
+    exprPath
+        .isCons(TImplTraitTypeParameter(_, _),
+          any(TypePath path0 | path0.isCons(getFutureOutputTypeParameter(), path)))
+    or
+    path = exprPath and
+    not (
+      exprPath = TypePath::singleton(TImplTraitTypeParameter(_, _)) and
+      result.(TraitType).getTrait() instanceof FutureTrait
+    ) and
+    not exprPath
+        .isCons(TImplTraitTypeParameter(_, _),
+          any(TypePath path0 | path0.isCons(getFutureOutputTypeParameter(), _)))
+  )
```

Comment:
Is it correct that this covers most of the special cases that occurs in practice but doesn't handle the general case where the awaited expression is some arbitrary type that implements `Future`?

It might be nice to add some comments about which cases are covered and about what is not covered.

For full generality I guess that we might have to expose `conditionSatisfiesConstraintTypeAt` in some form or implement `await` using the `Matching` module?

Reply:
I have reimplemented this.